ADR/0033-product-version-controller.md

```diff
@@ -0,0 +1,493 @@
+# 0032 Product and Version Controller
+
+Date: 2024-02-01
+
+## Status
+
+Proposed.
+
+## Context
+
+Konflux began its way as "App Studio", which was mainly designed to facilitate
+the development of online managed services. In order to do that a few major
+abstractions such as "Applications" and "Components" were introduced to help
+managing the underlying Tekton Pipeline and other lower level objects.
+
+With Konflux shifting focus to serving enterprise development teams requests
+had been made to enable using other abstraction metaphors such as "Product" and
+"Version" for managing development, build, test and release workflows. As one
+user put it:
+
+> As an architect, I need to be able to see all my apps versions in
+> one place, so that they are easily accessible by everyone on the
+> team and organized in a way that matches the way our team talks
+> about them, making my workflow for each app more streamlined.
+
+Some Konflux users have already attempted managing large multi-versioned
+products in Konflux. The pattern that emerged for doing so is as follows:
+
+* Create a separate Konflux Application for each product version
+* Use branches to track different source code versions within Git repositories
+* Use a separate Konflux Components for each buildable piece of code
+  (E.g. Dockerfile) in each branch in each repo that is included in a particular
+  application version.
+* To manage large amounts of components, some users have used multiple
+  application objects for a single product version.
+
+As one can expect, creating and managing so many Konflux objects is currently
+rather cumbersome. There are a few redundant steps that need to be taken. For
+example, access credentials for a Git repository need to be filled-in every time
+a component referring to that repository is created, even if other components
+referring that repository already exist in the user's workspace.
+
+The intention for the solution proposed in this document is to follow the
+pattern described above as far as the Konflux "Application" and "Component"
+objects go and provide a higher-level layer for the streamlined management of
+large amounts of such objects.
+
+An alternative approach to managing the challenges we have described above is to
+modify the way Application and Component object function so that for example:
```

Comment:
The ADR (https://github.com/redhat-appstudio/architecture/pull/167) that I submitted can be viewed as an alternative approach as well. By getting rid of our Component and Application model, we have the ability to remove many of the restrictions of the current system.

Reply:
The two might be complementary - "Artifact" and "ArtifactCollection" seem to be very abstract concepts, the controller described here could be used to link them to the terminology referred by the development teams.